iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/errors.hpp

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2023 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_HOOFS_ERROR_REPORTING_ERROR_HPP
+#define IOX_HOOFS_ERROR_REPORTING_ERROR_HPP
+
+#include <utility>
+
+#include "iceoryx_hoofs/error_reporting/types.hpp"
+
+namespace iox
+{
+namespace err
+{
+
+// We expect an error to have the following interface
+// 1. ErrorCode code() const
+// 2. ModuleId module() const
+
+// By default, there are only error codes and violations.
+// Custom errors can be added but must satisfy the minimal interface.
+
+class Violation
+{
+  public:
+    explicit Violation(ErrorCode code)
+        : m_code(code)
+    {
+    }
+
+    Violation(ErrorCode code, ModuleId module)
+        : m_code(code)
+        , m_module(module)
+    {
+    }
+
+    ErrorCode code() const
+    {
+        return m_code;
+    }
+
+    ModuleId module() const
+    {
+        return m_module;
+    }
+
+    bool operator==(const Violation& rhs) const
+    {
+        return m_code == rhs.m_code && m_module == rhs.m_module;
+    }
+
+    bool operator!=(const Violation& rhs) const
+    {
+        return !(*this == rhs);
+    }
+
+    static Violation createPreconditionViolation()
+    {
+        return Violation(ErrorCode(ErrorCode::PRECONDITION_VIOLATION));
+    }
+
+    static Violation createAssumptionViolation()
+    {
+        return Violation(ErrorCode(ErrorCode::ASSUMPTION_VIOLATION));
+    }
+
+  private:
+    ErrorCode m_code;
+    ModuleId m_module{ModuleId::ANY};
+};
+
+// primary template is the identity
+// this can be overriden by modules to handle specific errors
+template <typename ErrorLike>
+auto toError(ErrorLike&& value)
+{
+    return std::forward<ErrorLike>(value);
+}
+
+template <class Error>
+inline ErrorCode toCode(const Error& error)
+{
+    return error.code();
+}
+
+template <>
+inline ErrorCode toCode<ErrorCode>(const ErrorCode& error)
+{
+    return error;
+}
```

Comment:
I guess the purpose of this specialization is template meta-programming?

Reply:
It is needed as every error type (and the ErrorCode is the most simple one) must define a conversion to the ErrorCode (by overload/specialization). Otherwise the code does not compile.

Reply:
In a way this is all "concept" based but without using concepts, because we cannot for multiple reasons: C++ standard, here we actually prefer to have a free function to allow primitive types (such as `ErrorCode` as a thin wrapper) etc.